### ISSUE TITLE
LendingIterator with Rust 1.65+

### ISSUE THREAD

**Issue by fkohlgrueber** [danielhenrymantilla/lending-iterator.rs#10]
If I understand correctly, the release of Rust 1.65 should make it possible to implement lending iterators without workarounds. The [release notes](https://blog.rust-lang.org/2022/11/03/Rust-1.65.0.html#generic-associated-types-gats) even use lending iterators as an example of what's possible with GATs: 

```
/// An `Iterator`-like trait that can borrow from `Self`
trait LendingIterator {
    type Item<'a> where Self: 'a;

    fn next<'a>(&'a mut self) -> Option<Self::Item<'a>>;
}
```

I've been using the code above in a library I'm writing and it's working fine for my use-cases. I don't like defining the trait in my lib though and would prefer using a crate that would also implement common iterator methods for the lending iterator trait. 

My question now is whether this project will take advantage of GATs being stable now. For interoperability, it'd make sense to have a common library for lending iterators instead of each library defining their own version IMO. It's possibly even something worth including in the stdlib, but I didn't find discussions about it.

What do you think about this situation and what would you recommend?

**Comment by danielhenrymantilla** [comment]
Yeah, I intend to spend some time exploring usage of GATs over `nougat` in some cases (I might add a module for this, I think, with a `LendingIteratorGat`, for instance); but I haven't got the time for it as of yet.

Until then, I did post over Zulip a list of all the limitations Rust currently has for a clean (no workarounds) `LendingIterator` design:

  - https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/lib.20after.20GATs/near/307867488

**Comment by Crazytieguy** [comment]
@danielhenrymantilla Hi 👋
I spent some time working on a version of lending iterators that uses gats, doesn't need the higher kinded types and have an over all cleaner API IMO. Instead of publishing it as a separate crate I wanted to check if you would want to modify this crate. Ideally I'd like to work together to create what will become the canonical lending iterator 😄

https://github.com/Crazytieguy/lending-iterator

**Comment by danielhenrymantilla** [comment]
So, I am open to adding a `gat` module for a GAT-based implementation, such as @Crazytieguy (feel free to open a PR doing that 🙏). I won't be replacing the main API until at the very least something like:

```rs
let index: usize = …;
some_lending_iterator.map(|x: &mut [u8]| &mut x[index])
```

is capable of working / resulting in another fully usable lending iterator.

AFAIK, right now the only way to work with `.map()` / `.and_then` and the like is by:

  - either using free / non-closure `fn`s for which we can write an explicit lifetime-generic signature;
  - or to somehow annotate the closure itself with such a signature,
      - either on `nightly` through `closure_binders`:
        ```rs
        some_lending_iterator.map(for<'a> |x: &'a mut [u8]| -> &'a mut u8 { &mut x[index] })
        ```
      - or on stable through some helper macro which we could reëxport such as https://docs.rs/higher-order-closure:
        ```rs
        some_lending_iterator.map(hrtb!(|x: &mut [u8]| -> &mut u8 { &mut x[index] }))
        ```

In that regard, the current API of the crate using:

```rs
some_lending_iterator.map::<HKT!(&mut u8), _>(|[], x| &mut x[index])
```

or even the dedicated pervasive-case helper:

```rs
some_lending_iterator.map_mut(|x| &mut x[index])
```

still seem more convenient.

  - (for those unconvinced about `HKT!` _vs._ `hrtb!`, the former works with inference, whereas the latter does not, so when `u8` is, instead, some very complex type or a type which happens to mention generic parameters in scope, both `hrtb!` and `closure_binders` become quite painful to use; `HKT!`, on the other hand, lets you infer any part of the type that does not involve the higher order lifetime: `HKT!(&mut _)` would work, for instance (and `.map_mut` also Just Works™)).

**Comment by danielhenrymantilla** [comment]
For instance, something actually useful from having a `gat` module, on the other hand, would be for manual `LendingIterator` implementations:

```rs
struct MyLendingIter { … }

impl LendingIteratorGat for MyLendingIter {
    type GatItem<'next> = …;
    where
        Self : 'next,
    ;

    fn gat_next(this: &mut self) -> Option<Self::GatItem<'_>> {
        …
    }
}
```

and then magically having `MyLendingIter : LendingIterator` thanks to some blanket impl along the lines of:

```rs
#[gat]
impl<T : LendingIteratorGat> LendingIterator for T {
    type Item<'next> = T::GatItem<'next>;
    where
        Self : 'next,
    ;

    fn next(&mut self) -> Option<Self::Item<'_>> {
        T::gat_next(self)
    }
}
```

**Comment by Crazytieguy** [comment]
I see your point regarding the missing type inference, I agree `HKT!` is more powerful and we shouldn't settle for anything less powerful.

At this point, it's fairly obvious what the API will be in the long term, and it's possible to create a future-proof API today. I think having several "pervasive-case" helpers is fine, if we name them well we'll be able to deprecate them when `map` can handle all cases properly - and just not have `map` at all until then.

Gaining real users is valuable, and we can't gain many users before the API is simplified and future-proofed (people will want stability).

I can't see how adding a gat module will be future-proof.

Maybe consider it once more after all?

**Comment by WanderLanz** [comment]
Hi I'm quite the amateur here, but if I understand correctly from [Sabrina Jewson's Blog](https://sabrinajewson.org/blog/the-better-alternative-to-lifetime-gats), the recently stabilized GATs require 'static because HRTBs are not implicitly restricted by the GAT constraint and don't *lend* themselves to a lot of use cases. 😄

If I am not misunderstanding things here, the problem is type inference on HRTB closures and unstable inference with `Fn` + HRTB constraints, which we would still need something like [higher-order-closure](https://docs.rs/higher-order-closure) until HRTB closures are stabilized or an advancement in type inference and late-bound vs early-bound.

For example, Higher-Ranked lifetime on `Fn` trait Output type: 
```rust
for<'all> FnMut(<Self as LendingIter<'all>>::Item) -> <Out as HKT<'all>>::This
```
fails to compile because of late-bound shenanigans and a previous ICE requires a hard-coded fail?

From playing around with this problem myself, I haven't actually even used GATs at all, but I'm a bad programmer so maybe it presents itself in a place I haven't seen yet.

**Comment by WanderLanz** [comment]
Sorry if I'm being an idiot, 

I was using Lending Iterators as Rust practice (https://github.com/WanderLanz/Lender) and I was under the impression that GATs were not going to be a full solution to this problem from [Sabrina Jewson's blog](https://sabrinajewson.org/blog/the-better-alternative-to-lifetime-gats) and trying to use GATs myself, because the second you need a HRTB they can't implicitly constrain it.  
```rust
for<'all /* where Self: 'all */>
```

Is there already a proposal for qualified HRTB syntax `for<'all where Self: 'all>` or will this just be for the most basic APIs (e.g. `next`)?

**Comment by WanderLanz** [comment]
Well, from my experience trying to do this with just implictly-bounded HRTBs and avoiding HKTs [(r)](https://github.com/WanderLanz/Lender), there's definitely alot to say about how ergonomic a GAT would be, especially with closures.

I don't think it needs to be a module if it's already a separate crate, and in fact you could probably take some of the API out that use HRTBs just to avoid confusion when the struct or data isn't explicitly 'static, being an ecosystem-friendly version of the DIY-trait with some QOL stuffs like `count` that don't require HRTBs. Most of HRTB stuff is still doable imperatively anyways.

**Comment by Crazytieguy** [comment]
> I don't think it needs to be a module if it's already a separate crate

I don't follow, is there a separate crate? I don't understand if you're suggesting creating another crate without HKTs or if you're suggesting something for this crate.

**Comment by WanderLanz** [comment]
Oh sorry I was referencing the example GAT implementation that you made. Not a lot of sleep when writing my comment, sorry it wasn't clear. Could definitely be a good module/feature of this crate if it ends up working well with generic implementations, but my gist was that it has plenty of merit as a standalone crate (a GAT lending-iterator interface) if not just to standardize a GAT implentation.
